python/mxnet/text/glossary.py

```diff
@@ -0,0 +1,654 @@
+                'load. Call ' \
+                'mxnet.text.Embedding.show_embed_names_and_pretrain_files() ' \
+                'to see the available embed_name and pretrain_file.'
+
+        self._special_init_vec = special_init_vec
+
+        # User specifies pretrained embedding file at the path of pretrain_file.
+        if os.path.isfile(pretrain_file):
+            pretrain_file_path = pretrain_file
+        else:
+            pretrain_file_path = \
+                Embedding._get_pretrain_file_path_from_url(pretrain_file, url,
+                                                           embed_name,
+                                                           embed_root)
+
+        self._load_embedding(pretrain_file_path, token_delim)
+
+        assert len(self._idx_to_vec) - 1 == len(self._idx_to_token) \
+            == len(self._token_to_idx), \
+            'The extra (last) row of self._idx_to_vec is the initialized ' \
+            'embedding vector for a special, such as an unknown token.'
+
+    @staticmethod
+    def _get_pretrain_file_path_from_url(pretrain_file, url, embed_name,
+                                         embed_root):
+        """Get the local path to the pre-trained file from the url.
+
+        The pretrained embedding file will be downloaded from url if it has not
+        been downloaded yet or the existing file fails to match its expected
+        SHA-1 hash.
+        """
+        embed_dir = os.path.join(embed_root, embed_name)
+        pretrain_file_path = os.path.join(embed_dir, pretrain_file)
+        download_file = os.path.basename(url)
+        download_file_path = os.path.join(embed_dir, download_file)
+
+        embed_cls = Embedding.embed_registry[embed_name]
+        expected_file_hash = embed_cls.pretrain_file_sha1[pretrain_file]
+
+        if hasattr(embed_cls, 'pretrain_archive_sha1'):
+            expected_download_hash = \
+                embed_cls.pretrain_archive_sha1[download_file]
+        else:
+            expected_download_hash = expected_file_hash
+
+        # The specified pretrained embedding file does not exist or fails to
+        # match its expected SHA-1 hash.
+        if not os.path.isfile(pretrain_file_path) or \
+                not check_sha1(pretrain_file_path, expected_file_hash):
+            # If download_file_path exists and matches
+            # expected_download_hash, there is no need to download.
+            download(url, download_file_path,
+                     sha1_hash=expected_download_hash)
+
+        # If the downloaded file does not match its expected SHA-1 hash,
+        # we do not encourage to load embeddings from it in case that its
+        # data format is changed.
+        assert check_sha1(download_file_path, expected_download_hash), \
+            'The downloaded file %s does not match its expected SHA-1 ' \
+            'hash. This is caused by the changes at the externally ' \
```

Comment:
“This is caused” -> “This is likely caused” since failed/partial/interrupted download can also cause this.

Reply:
Thanks. The immediately previous if-block makes sure that it can only caused by external changes.